Infrastructure followup

nstarman opened the pull request.

patrick-kidger commented on pyproject.toml:
```diff
@@ -65,3 +65,9 @@ include = ["quax", "tests"]
 dev = [
     "pre-commit>=4.1.0",
 ]
+test = [
```
I am also only just realising that we have `[dependency-groups]` here! I missed that when it came in.

It's a minor point but PEP 735 is still pretty new so I think I'd prefer to keep these as `[project.optional-dependencies]`.

nstarman replied:
Migrating.

nstarman replied:
Done.

nstarman commented on .github/workflows/run_tests.yml:
Done. The only `uv` I've kept is for the oldest supported dependencies test, which is significantly easier with `uv` than with pip.

nstarman commented:
@patrick-kidger I think I've addressed your comments. I've un-`uv`-ified the CI (except for the new oldest-dependencies check) and moved dependency groups to optional dependencies so that pip can see them.

nstarman commented:
@patrick-kidger the fox for Jax v0.5.3 is also ready.

patrick-kidger commented on .github/workflows/run_tests.yml:
```diff
@@ -9,7 +9,23 @@ env:
   FORCE_COLOR: 3
 
 jobs:
+  format:
```
Q: what's the rationale for making this a separate job / what goes wrong with keeping it in one job?

nstarman replied:
I had a few motivations.
1. Separation of concerns
2. Fast failure when there's a problem. Pre-commit doesn't require the package to be installed and caches the environment, so it spins up faster and fails fast if there's an issue.
3. Minimizing CI usage. Instead of failing 4 jobs when there's a linting error, only one job fails.
4. There's now 2 types of CI tests, the standard and the oldest-supported. Now all tests start with the pre-commit.

patrick-kidger replied:
This makes sense to me. I like the justification.

patrick-kidger commented on .github/workflows/run_tests.yml:
```diff
@@ -28,12 +44,33 @@ jobs:
+          python-version: "3.10"
+
+      - name: Install the project
+        run: uv sync --extra test --resolution lowest
```
Q: When using `uv` I've typically used `--resolution lowest-direct`, because typically there is some transitive dependency which hasn't correctly specified its lower bounds. No strong feelings on it here -- we have almost zero dependencies anyway -- but I'm just curious about your experience on this one?

nstarman replied:
Yeah, I've used both and found ``--resolution lowest-direct`` to be problematic.
The issue is that `lowest-direct` only does the earliest version of the direct dependency but then the most-up-to-date version of all other dependencies, which is not a realistic reflection of a user's environment. ```--resolution lowest`` is much more realistic. It does force the slightly annoying requirement of setting constraints in uv (as done here) when dependencies don't have proper minimum version pins (e.g. Jupyter is bad at this), but for me this has repeatedly proven useful in diagnosing problems.

patrick-kidger commented on pyproject.toml:
```diff
@@ -30,6 +30,17 @@ dependencies = [
 ]
 
+[project.optional-dependencies]
+dev = [
+    "pre-commit>=4.1.0",
+]
+test = [
```
Nit: let's merge these two groups into just one `dev` dependency, I don't htink there's a usecase for keeping the two separate?

nstarman replied:
The difference is how many tools, e.g. uv and pixi, treat `dev` versus other dependencies. When doing development they install `dev` automatically but not other extras.

patrick-kidger replied:
Okay. I'm familiar with the special-casing for dev. FWIW I think these probably should still be merged (I'll do this in a follow-up PR) as the test dependencies here are lightweight - but I can see in general that might be a useful separation if the test dependencies are heavy and one prefers to rely on CI instead.

patrick-kidger commented on tests/requirements.txt:
Nit: I think we can remove this file now?

patrick-kidger replied:
Yup, I need to upgrade the deployment script. That's on me though. I'll do the update and remove this file in a follow up PR.

patrick-kidger reviewed (commented):
Okay, this LGTM! Thank you for the updates.
I've left two minor nits + two questions for my own understanding. After those then we'll get this merged.

patrick-kidger reviewed (commented):
Alright! Thanks for answering my questions, and merged :)

nstarman commented:
Great! Thanks for the fast followup.
